File: havsim/simulation/simulation-cpp/test_set_lc.cpp
#include "vehicle.h"
#include "lane.h"
#include <iostream>
#include <cassert>
#include <fstream>

using namespace std;

void set_state(Vehicle *veh, double pos, double speed) {
    veh->set_pos(pos);
    veh->set_speed(speed);
}

void verify_output(const string &expected_output_path, const string &output) {
    ifstream test_file(expected_output_path);
    string line;
    if (test_file.is_open()) {
        getline(test_file, line);
        assert(line == output);
    } else {
        cerr << "Unable to open file";
    }
}

vector<vector<vector<double>>> read_input(const string &input_path) {

    const int x = 2000;
    const int y = 7;
    const int z = 2;
    vector<vector<vector<double>>> res(x, vector<vector<double>>(y, vector<double>(x, 0.0)));
    size_t i = 0;
    size_t j = 0;
    size_t k = 0;
    ifstream test_file(input_path);
    string line;
    if (test_file.is_open()) {
        getline(test_file, line);
        size_t pos;
        while ((pos = line.find(',')) != string::npos) {

            res[i][j][k] = stod(line.substr(0, pos));

            line.erase(0, pos + 1);
            k++;
            if (k == z) {
                k = 0;
                j++;
                if (j == y) {
                    j = 0;
                    i++;
                }
            }
        }
        res[i][j][k] = stod(line);
        test_file.close();
    } else {
        cerr << "Unable to open file";
    }
    return res;

}

char maketest(vector<Vehicle *> &vehlist, vector<Vehicle *> &vehlist2, vector<vector<double>> &data, Vehicle *testveh) {
    for (int i = 0; i < vehlist.size(); i++) {
        set_state(vehlist[i], data[i][0], data[i][1]);
    }

    for (auto veh:vehlist2) {
        veh->set_headway(veh->get_lane()->get_headway(veh, veh->get_lead()));
        veh->set_cf(0, 1);
    }
    unordered_map<Vehicle *, std::string> lc_actions;
    testveh->set_lc(lc_actions, 0, 1);
    if (lc_actions.count(testveh)) {
        return lc_actions[testveh][0];
    }

    return 'n';
}

int main() {
    Lane lane0({});
    Lane lane1({});
    Lane lane2({});
    Vehicle testveh(0, &lane1, {27, 1.2, 2, 1.1, 1.5}, {-2, .2, .2, 0, .1, 1});
    Vehicle lead(0, &lane1, {28, 1.2, 2, 1.1, 1.5}, {-1.5, .2, .2, 0, .1, 1});
    Vehicle fol(0, &lane1, {25, 1.2, 2, 1.1, 1.5}, {-1.5, .2, .2, 0, .1, 1});
    Vehicle lfol(0, &lane0, {20, 1.2, 1.1, 1.1, 1.5}, {-1.5, .2, .2, 0, .1, 1});
    Vehicle llead(0, &lane0, {23, 1.2, 2, 1.1, 1.5}, {-1.5, .2, .2, 0, .1, 1});
    Vehicle rfol(0, &lane2, {33, .9, 1.5, 1.1, 1.5}, {-1.5, .2, .2, 0, .1, 1});
    Vehicle rlead(0, &lane2, {27, 1.2, 2, 1.1, 1.5}, {-1.5, .2, .2, 0, .1, 1});
    testveh.set_llane(&lane0);
    testveh.set_rlane(&lane2);
    testveh.set_lead(&lead);
    testveh.set_fol(&fol);
    testveh.set_lfol(&lfol);
    testveh.set_rfol(&rfol);
    lead.set_fol(&testveh);
    fol.set_lead(&testveh);
    rfol.set_lead(&rlead);
    lfol.set_lead(&llead);

    // test1
    auto test1_input = read_input("../test1_input");
    testveh.set_l("discretionary");
    testveh.set_r("discretionary");
    vector<Vehicle *> vehlist = {&testveh, &llead, &lead, &rlead, &lfol, &fol, &rfol};
    vector<Vehicle *> vehlist2 = {&testveh, &lfol, &fol, &rfol};
    string test1_output;
    for (auto &i:test1_input) {
        test1_output += maketest(vehlist, vehlist2, i, &testveh);
    }
    verify_output("../test1_output", test1_output);


    // test2
    auto test2_input = read_input("../test2_input");
    testveh.set_l("discretionary");
    testveh.set_r("mandatory");

    string test2_output;
    for (auto &i:test2_input) {
        test2_output += maketest(vehlist, vehlist2, i, &testveh);
    }
    verify_output("../test2_output", test2_output);

    // test3
    auto test3_input = read_input("../test3_input");
    testveh.set_l("discretionary");
    testveh.set_r("");

    string test3_output;
    for (auto &i:test3_input) {
        test3_output += maketest(vehlist, vehlist2, i, &testveh);
    }
    verify_output("../test3_output", test3_output);

    // test4
    // same input as test1
    auto test4_input = read_input("../test1_input");
    testveh.set_in_relax(true);
    testveh.set_relax({20});
    testveh.set_relax_start(0);
    rfol.set_in_relax(true);
    rfol.set_relax({-30});
    rfol.set_relax_start(0);
    testveh.set_l("discretionary");
    testveh.set_r("discretionary");
    string test4_output;
    for (auto &i:test4_input) {
        test4_output += maketest(vehlist, vehlist2, i, &testveh);
    }
    verify_output("../test4_output", test4_output);


}

File: havsim/simulation/simulation-cpp/test_set_lc.py
"""
@author: rlk268@cornell.edu
"""
import numpy as np
# tests for set_lc method - turn off cooperative model
from havsim.simulation.simulation import lane, vehicle

# make dummy road network
road = {'name': 'test'}
lane0 = lane(0, 1000, road, 0)
lane1 = lane(0, 1000, road, 1)
lane2 = lane(0, 1000, road, 2)

# make vehicles for test
testveh = vehicle(0, lane1, [27, 1.2, 2, 1.1, 1.5], [-2, .2, .2, 0, .1, 1])
lead = vehicle(0, lane1, [28, 1.2, 2, 1.1, 1.5], [-1.5, .2, .2, 0, .1, 1])
fol = vehicle(0, lane1, [25, 1.2, 2, 1.1, 1.5], [-1.5, .2, .2, 0, .1, 1])
lfol = vehicle(0, lane0, [20, 1.2, 1.1, 1.1, 1.5], [-1.5, .2, .2, 0, .1, 1])
llead = vehicle(0, lane0, [23, 1.2, 2, 1.1, 1.5], [-1.5, .2, .2, 0, .1, 1])
rfol = vehicle(0, lane2, [33, .9, 1.5, 1.1, 1.5], [-1.5, .2, .2, 0, .1, 1])
rlead = vehicle(0, lane2, [27, 1.2, 2, 1.1, 1.5], [-1.5, .2, .2, 0, .1, 1])
# set relationships needed to test set lc
testveh.llane = lane0
testveh.rlane = lane2
testveh.lead = lead
testveh.fol = fol
testveh.lfol = lfol
testveh.rfol = rfol
lead.fol = testveh
fol.lead = testveh
rfol.lead = rlead
lfol.lead = llead


# helper fns
def set_state(veh, pos, spd):
    veh.pos = pos
    veh.speed = spd


def maketest(curdata):
    vehlist = [testveh, llead, lead, rlead, lfol, fol, rfol]
    vehlist2 = [testveh, lfol, fol, rfol]
    for count, veh in enumerate(vehlist):
        set_state(veh, curdata[count, 0], curdata[count, 1])
    for veh in vehlist2:
        veh.hd = veh.lane.get_headway(veh, veh.lead)
        veh.set_cf(0, 1)
    lc_actions = {}
    testveh.set_lc(lc_actions, 0, 1)
    return lc_actions[testveh] if testveh in lc_actions else 'n'


def dump_input(data, file_name):
    flat = data.flatten()
    formatted = ','.join([str(i) for i in flat])

    with open(file_name, 'w') as f:
        f.write(formatted)


def dump_output(data, file_name):
    formatted = ''.join(data)
    with open(file_name, 'w') as f:
        f.write(formatted)


# test 1- discretionary both sides
# other tests are basically just copy paste of this
# generate test data
np.random.seed(1)
test1 = np.random.rand(2000, 7, 2)
for i in range(len(test1)):
    curdata = test1[i]
    curdata[:, 0] = curdata[:, 0] * 75
    curdata[:, 1] = curdata[:, 1] * 25
    curdata[1:4, 0] += curdata[0, 0]
    curdata[4:, 0] -= curdata[0, 0]
dump_input(test1, 'test1_input')

testveh.l = 'discretionary'
testveh.r = 'discretionary'
test1result = [maketest(i) for i in test1]
dump_output(test1result, 'test1_output')

# test 2 - mandatory one side
np.random.seed(2)
test2 = np.random.rand(2000, 7, 2)
for i in range(len(test2)):
    curdata = test2[i]
    curdata[:, 0] = curdata[:, 0] * 40
    curdata[:, 1] = curdata[:, 1] * 15
    curdata[1:4, 0] += curdata[0, 0]
    curdata[4:, 0] -= curdata[0, 0]
dump_input(test2, 'test2_input')

testveh.l = 'discretionary'
testveh.r = 'mandatory'
test2result = [maketest(i) for i in test2]
dump_output(test2result, 'test2_output')

np.random.seed(3)
# test 3 - discretionary one side
test3 = np.random.rand(2000, 7, 2)
for i in range(len(test3)):
    curdata = test3[i]
    curdata[:, 0] = curdata[:, 0] * 75
    curdata[:, 1] = curdata[:, 1] * 25
    curdata[1:4, 0] += curdata[0, 0]
    curdata[4:, 0] -= curdata[0, 0]
dump_input(test3, 'test3_input')
testveh.l = 'discretionary'
testveh.r = None
test3result = [maketest(i) for i in test3]
dump_output(test3result, 'test3_output')

# test 4 - relaxation (turn relax settings to True - True)
testveh.in_relax = True
testveh.relax = [20]
testveh.relax_start = 0
rfol.in_relax = True
rfol.relax = [-30]
rfol.relax_start = 0

test4 = test1

testveh.l = 'discretionary'
testveh.r = 'discretionary'
test4result = [maketest(i) for i in test4]
dump_output(test4result, 'test4_output')
